src/sink.rs

```diff
@@ -0,0 +1,130 @@
+//! Rate-limiting combinator for sinks.
+
+use crate::Ratelimit;
+use futures::{Async, AsyncSink, Future, Poll, Sink, StartSend, Stream};
+use ratelimit_meter::{algorithms::Algorithm, DirectRateLimiter, NonConformance};
+use std::io;
+
+pub trait SinkExt: Sink {
+    /// Limits the rate at which items can be put into the current sink.
+    ///
+    /// Note that this combinator limits the rate at which it accepts items,
+    /// not necessarily the rate at which the underlying sink does.
+    /// The combinator will buffer at most one item in order to adhere to the
+    /// given limiter. I.e. if it already has an item buffered and needs to
+    /// wait, it will not accept any more items until the underlying stream has
+    /// accepted the current item.
+    fn sink_ratelimit<A: Algorithm>(self, limiter: DirectRateLimiter<A>) -> Ratelimited<Self, A>
```

Comment:
I think this might scan better in user code if it was named `ratelimit_send`.

Reply:
I'm not sure I like having `send` in the name because it doesn't actually send anything like all the other `*send*` methods in the `Sink` trait do (it only creates a combinator).
And then it looks weird when you do actually send something:
`my_sink.ratelimit_send(…).send_all(…)`

Not that `sink_ratelimit` would look much better...
Personally I'd like it to just be `ratelimit` but that clashes with the method for streams which is why I went with the same `sink_` prefix the futures crate uses for `map_err` and `from_err`.

Ultimately I don't mind either way, you decide.

Reply:
Ah, using the same convention as the futures crate sounds good. OK!